quinn/src/platform/unix.rs

```diff
@@ -165,21 +167,52 @@ impl super::UdpExt for UdpSocket {
         Ok(sent)
     }
 
-    fn recv_ext(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr, Option<EcnCodepoint>)> {
-        let mut name = MaybeUninit::<libc::sockaddr_storage>::uninit();
-        let mut iov = libc::iovec {
-            iov_base: buf.as_ptr() as *mut _,
-            iov_len: buf.len(),
+    #[cfg(not(any(target_os = "macos", target_os = "ios")))]
+    fn recv_ext(&self, bufs: &mut [IoSliceMut<'_>], meta: &mut [RecvMeta]) -> io::Result<usize> {
+        use crate::udp::BATCH_SIZE;
+        let mut names = [MaybeUninit::<libc::sockaddr_storage>::uninit(); BATCH_SIZE];
+        let mut ctrls = [cmsg::Aligned(MaybeUninit::<[u8; CMSG_LEN]>::uninit()); BATCH_SIZE];
+        let mut hdrs = unsafe { mem::zeroed::<[libc::mmsghdr; BATCH_SIZE]>() };
+        let max_msg_count = bufs.len().min(BATCH_SIZE);
+        for i in 0..max_msg_count {
+            prepare_recv(
+                &mut bufs[i],
+                &mut names[i],
+                &mut ctrls[i],
+                &mut hdrs[i].msg_hdr,
+            );
+        }
+        let msg_count = loop {
+            let n = unsafe {
+                libc::recvmmsg(
+                    self.as_raw_fd(),
+                    hdrs.as_mut_ptr(),
+                    bufs.len().min(BATCH_SIZE) as libc::c_uint,
+                    0,
+                    ptr::null_mut(),
+                )
+            };
+            if n == -1 {
+                let e = io::Error::last_os_error();
+                if e.kind() == io::ErrorKind::Interrupted {
+                    continue;
+                }
+                return Err(e);
+            }
+            break n;
```

Comment:
Want to add an `assert!(n >= 0)` here, to catch other weird OS return values?

Reply:
We're using `== -1` for various syscalls throughout this file. If there's reason to believe that might not be portable for any of them (is there?) we should probably address it for them all in a separate PR.

Reply:
It should be portable, and the doc say it's `-1`. However I sometimes feel like not trusting foreign code :)